### PR TITLE
feat: add jest

### PR DESCRIPTION
**Pull request by ncovercash**
what it says on the tin

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name: Lint, build, and test
```
Comment (ncovercash) [inline .github/workflows/ci.yaml:1]: this checks things and nags you if a commit breaks everything (email and/or PR check and/or as a check/x on the commit)

File: jest.config.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+/** @type {import('ts-jest').JestConfigWithTsJest} */
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'jsdom',
+
+  testMatch: ['**/src/**/?(*.)test.{js,jsx,ts,tsx}'],
+  setupFilesAfterEnv: ['./src/test/jest.setup.ts'],
+
+  transform: {
+    '^.+\\.(js|jsx|ts|tsx)$': ['ts-jest', { tsconfig: './tsconfig.test.json' }],
```
Comment (ncovercash) [inline jest.config.js:14]: we have to customize tsconfig here...see later for why

File: package.json [hunk 3]
```diff
@@ -55,7 +58,10 @@
   },
   "devDependencies": {
     "@auth/drizzle-adapter": "^0.3.11",
+    "@testing-library/jest-dom": "^6.2.0",
+    "@testing-library/react": "^14.1.2",
```
Comment (ncovercash) [inline package.json:62]: these do react

File: src/test/jest.setup.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+import '@testing-library/jest-dom';
```
Comment (ncovercash) [inline src/test/jest.setup.ts:1]: this is run as part of every test, effectively extending Jest's functionality with this. This allows asserting that things are in a DOM, etc

File: src/testTest.test.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+it("works", () => {
+  expect(true).toBe(true);
+})
```
Comment (ncovercash) [inline src/testTest.test.ts:3]: basic plain/non-react test. there's [tons of "matchers"](https://jestjs.io/docs/using-matchers), such as `toBe`, `toEqual`, `toBeUndefined`, etc etc)

File: src/testWithReact.test.tsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { render, screen } from '@testing-library/react';
+
+it('likes react, too!', () => {
+  render(
+    <p>foo bar</p>
+  );
+
+  expect(screen.getByText('foo bar')).toBeInTheDocument();
```
Comment (ncovercash) [inline src/testWithReact.test.tsx:8]: The general workflow for react tests is:

1. render the element
2. get the element. preferred way to do this for accessibility is via `screen.getByRole` where we ask it for something like a `heading`, `textbox`, etc (ensures screenreaders/etc controls can find the element, too)
3. expect it to be whatever, in the document, have some content, be enabled/disabled, anything you can imagine

File: tsconfig.test.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "jsx": "react-jsx"
```
Comment (ncovercash) [inline tsconfig.test.json:4]: next uses a different option that is incompatible with Jest

**Comment by kalkafox** [comment]
Thanks for the help!

**Pull request merged**